### PR TITLE
SSD1309 Implementation

### LINKED ISSUE #222: Support for SSD1309
Hi,
I just wonder if any plan to support SSD1309 LCD? Thank you.

### PR DESCRIPTION
**Pull request by rm-hull**
Fixes #222 

## TODO
- [x] Driver code
- [x] Add documentation
- [x] Add PDF datasheet
- [x] Regression tests
- [ ] Add `ssd1309.conf` to luma.examples
- [x] Add entry to changelog
- [x] Ready for review

https://www.aliexpress.com/item/2-42-inch-LCD-Screen-12864-OLED-Display-Module-IIC-I2C-SPI-Serial-C51-STM32-SSD1309/32857123469.html?spm=a2g0s.9042311.0.0.3da24c4dpFOWfw

image

Device appears to be configured for SPI rather than I2C.

| OLED Pin | OLED Name | RPi Pin | RPi Function | Remarks |
|----------|----------|---------|----------|-------|
| 1 | GND | P01-20 | GND | Ground |
| 2 | VCC | P01-17 | 3V3 | +3.3V Power |
| 3 | SCL | P01-23 | GPIO 11 (SCLK) | Serial Clock |
| 4 | SDA | P01-19 | GPIO 10 (MOSI) | Serial Data |
| 5 | RES | P01-22 | GPIO 25 | Reset |
| 6 | DC | P01-18 | GPIO 24 | Data/Command |
| 7 | CS | P01-24 | GPIO 8 (CE0) | Chip Select |

### CODE REVIEW

**Comment by rm-hull** [comment]
Reusing the SSD1306 driver, with the following config settings:
```
$ cat conf/ssd1309.conf 
--display=ssd1306
--interface=spi
--width=128
--height=64
--spi-bus-speed=8000000
```

Gives:

image

Picture quality is a bit fuzzy from my phone, but the display appears to work with the SSD1306 driver. It doesn't look to handle brightness levels as is, so perhaps there are a few tweaks to be made.

I might just extend the SSD1306 driver into a new SSD1309 class with the requisite props to control brightness

File: luma/oled/device/__init__.py [hunk 1]
```diff
@@ -202,6 +202,17 @@ def display(self, image):
...
 
+class ssd1309(ssd1306):
+    """
+    Serial interface to a monochrome SSD1309 OLED display.
+
+    On creation, an initialization sequence is pumped to the display
+    to properly configure it. Further control commands can then be called to
+    affect the brightness and other settings.
+    """
+    pass
```
[the remark this answers is not in the document]
Comment (rm-hull) [reply luma/oled/device/__init__.py]: Amended

File: luma/oled/device/__init__.py [hunk 2]
```diff
@@ -202,6 +202,16 @@ def display(self, image):
...
 
 
+class ssd1309(ssd1306):
+    """
+    Serial interface to a monochrome SSD1309 OLED display.
+
+    On creation, an initialization sequence is pumped to the display
+    to properly configure it. Further control commands can then be called to
+    affect the brightness and other settings.
+    """
```
[the remark this answers is not in the document]
Comment (rm-hull) [reply luma/oled/device/__init__.py]: Also added docstring params for ssd1309 and ssd1306

[the remark this answers is not in the document]
Comment (thijstriemstra) [reply luma/oled/device/__init__.py]: are those documented params not automatically copied by sphinx and inherited from ssd1306?

[the remark this answers is not in the document]
Comment (rm-hull) [reply luma/oled/device/__init__.py]: Interestingly, the SSD1309 device doesnt even show up in the API docs: https://luma-oled.readthedocs.io/en/feature-ssd1309/api-documentation.html ...

[the remark this answers is not in the document]
Comment (thijstriemstra) [reply luma/oled/device/__init__.py]: It's present in class diagram but no API docs indeed, weird, i'll take a look. Ssd1309 can also be released later, what do you think?

**Comment by nvl1109** [comment]
@rm-hull Hi Richard,
Many thanks for your effort. I have tried this pull request on Raspberry Pi 0 W + SSD1309 display, first 20mins it works correctly, after that the screen suddenly become horizontal mirror.
My code just draw 5 lines of text, clock update (first line) every second.
Does it happen on your screen when using for long time or not? Thank you.

**Comment by thijstriemstra** [comment]
> My code just draw 5 lines of text, clock update (first line) every second.

@nvl1109 can you share that code?

**Comment by nvl1109** [comment]
> > My code just draw 5 lines of text, clock update (first line) every second.
> 
> @nvl1109 can you share that code?

Hi Thijs and all,
Here is my class for display: https://pastebin.com/1jWzxRdm
The initialization of display is in ``_setup()`` function.
The drawing code is in ``show()`` function.
Thank you.

File: luma/oled/device/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thijstriemstra) [reply luma/oled/device/__init__.py]: @rm-hull it was missing in `__all__`, fixed that now.

**Comment by rm-hull** [comment]
@nvl1109 I'll try and run your code against my SSD1309 and see what happens. Can you take a photo of what the screen looks like when you said _"after that the screen suddenly become horizontal mirror."_
Do you mean the characters are reversed?

Anyway, probably better to track this in a separate github issue rather than in this PR, if you wouldnt mind creating one please and attaching the photo.

File: luma/oled/device/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thijstriemstra) [reply luma/oled/device/__init__.py]: this does not seem to render properly:

```
serial_interface – the serial interface (usually a :py:class`luma.core.interface.serial.spi` instance) 
```

**Comment by nvl1109** [comment]
Here is the picture.
https://drive.google.com/file/d/1pwVVsCdoF37n66kC47qKMiyGy7wup7ps/view?usp=sharing

I put comment here because of the feature is not merged.

**Pull request merged**